src/relay/pass/pointer_analysis.cc

```diff
@@ -0,0 +1,168 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * Copyright (c) 2019 by Contributors
+ *
+ * \file partial_eval.cc
+ *
+ * \brief Pointer analysis via Andersen's Algorithm.
+ * Analyze may-alias, escape, pointer content.
+ */
+#include <tvm/relay/analysis.h>
+#include <tvm/relay/expr_functor.h>
+#include "pass_util.h"
+
+namespace tvm {
+namespace relay {
+
+struct AllocAbstractLocation : ExprVisitor {
+  Map<Expr, AbstractLocation> spawn;
+  Map<AbstractLocation, Expr> origin;
+  void VisitExpr_(const RefCreateNode* op) final {
+    Expr refc = GetRef<RefCreate>(op);
+    auto loc = AbstractLocationNode::make(spawn.size() + 1);  // 0 is reserved for external ref
+    spawn.Set(refc, loc);
+    origin.Set(loc, refc);
+    ExprVisitor::VisitExpr_(op);
+  }
+};
+
+const AbstractLocation external_ref = AbstractLocationNode::make(0);
+
+struct Analyze : ExprFunctor<void(const Expr&)> {
+  Map<Expr, AbstractLocation> spawn;
+  Map<AbstractLocation, Expr> origin;
+  Map<Expr, Set<AbstractLocation>> contain;
+  Map<AbstractLocation, Set<Expr>> store;
+  bool progress = true;
+  Analyze(Map<Expr, AbstractLocation> spawn,
+          Map<AbstractLocation, Expr> origin) : spawn(spawn), origin(origin) {
+    for (const auto& p : origin) {
+      store.Set(p.first, Set<Expr>());
+    }
+  }
+
+  void analyze(const Expr& e) {
+    progress = false;
+    VisitExpr(e);
+  }
+
+  void VisitExpr(const Expr& e) final {
+    if (contain.count(e) == 0) {
+      contain.Set(e, Set<AbstractLocation>());
+    }
+    ExprFunctor::VisitExpr(e);
+  }
+
+  void VisitExpr_(const ConstantNode* op) final { }
+  void VisitExpr_(const OpNode* op) final { }
+  void VisitExpr_(const ConstructorNode* op) final { }
+
+  void FlowTo(AbstractLocation from, const Expr& to) {
+    Set<AbstractLocation> locs = contain.at(to);
+    if (locs.count(from) == 0) {
+      progress = true;
+      locs.Insert(from);
+      contain.Set(to, locs);
+    }
+  }
+
+  void FlowTo(const Expr& from, AbstractLocation to) {
+    Set<Expr> exprs = store.at(to);
+    if (exprs.count(from) == 0) {
+      progress = true;
+      exprs.Insert(from);
+      store.Set(to, exprs);
+    }
+  }
+
+  void FlowTo(const Expr& from, const Expr& to) {
+    for (auto loc : contain.at(from)) {
+      FlowTo(loc, to);
+    }
+  }
+
+  void VisitExpr_(const CallNode* op) final {
+    VisitExpr(op->op);
+    for (Expr e : op->args) {
+      VisitExpr(e);
+    }
+    Expr expr = GetRef<Expr>(op);
+    CHECK(op->op.as<OpNode>()) << "Only support call to operator right now.";
```

Comment:
Should this really be a `CHECK` (enforcing an internal invariant), rather skipping it if not supported and issuing a warning?

It seems a little confusing that you visit the called function above (which would still work if it is not an operator) but have this check down here.

Reply:
@slyubomirsky ultimately, andersen's algorithm is an first order algorithm. It cannot be supported. It is indeed true that it will error before reaching here though. I am investigating using AAM to do higher order analysis, but it is very complex and this is a testing ground.

Reply:
Ah, that would be a good thing to document then